src/main/kotlin/com/demonwav/mcdev/creator/ProjectCreator.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Minecraft Dev for IntelliJ
+ *
+ * https://minecraftdev.org
+ *
+ * Copyright (c) 2020 minecraft-dev
+ *
+ * MIT License
+ */
+
+package com.demonwav.mcdev.creator
+
+import com.demonwav.mcdev.creator.buildsystem.BuildSystem
+import com.intellij.openapi.module.Module
+import java.nio.file.Path
+
+/**
+ * This class represents a specific configuration for a project to be created. Typically these configurations represent
+ * ([PlatformType][com.demonwav.mcdev.platform.PlatformType], [BuildSystem][com.demonwav.mcdev.creator.buildsystem.BuildSystem])
+ * pairs, but this API does not apply any such restriction. Each `BuildSystem` defines how to create a [ProjectCreator]
+ * instance for the given configuration using the [buildCreator][com.demonwav.mcdev.creator.buildsystem.BuildSystem.buildCreator]
+ * method.
+ *
+ * &nbsp;
+ *
+ * In the current implementation the creation of [ProjectCreator] instances are delegated to project configurations via
+ * implementing the [GradleCreator][com.demonwav.mcdev.creator.buildsystem.gradle.GradleCreator] and/or
+ * [MavenCreator][com.demonwav.mcdev.creator.buildsystem.maven.MavenCreator] interfaces. This allows some configurations
+ * to produce different [ProjectCreator] instances based on their specific configuration, such as
+ * [ForgeProjectConfig][com.demonwav.mcdev.platform.forge.creator.ForgeProjectConfig].
+ */
+interface ProjectCreator {
+
+    /**
+     * Returns the [CreatorStep]s which should be executed in order to create the project configuration represented by
+     * this [ProjectCreator].
+     */
+    fun getSingleModuleSteps(): Iterable<CreatorStep>
+
+    /**
+     * Returns the [CreatorStep]s which shoud be executed in order to create the submodule project configuration
+     * represented by this [ProjectCreator].
+     */
+    fun getMultiModuleSteps(projectBaseDir: Path): Iterable<CreatorStep>
+}
+
+abstract class BaseProjectCreator(
+    private val rootModule: Module,
+    private val buildSystem: BuildSystem,
+    private val config: ProjectConfig
+) : ProjectCreator {
+    protected val project
+        get() = rootModule.project
+
+    abstract fun getMainClassText(packageName: String, className: String): String
+
+    fun setupMainClassStep(): BasicMainClassStep {
+        val (packageName, className) = splitPackage(config.baseOrError.mainClass)
+        val mainClassText = getMainClassText(packageName, className)
+        return BasicMainClassStep(project, buildSystem, config.baseOrError, mainClassText)
+    }
```

Comment:
How would a good example Fabric mod be setup like from the plugin? What I mean is would you not generate a main class or would you generate multiple main classes? Or would that be dependent on the user's selections?

Reply:
Yeah, dependent on user selection. In my fork, the entrypoints UI looks like this:
![image](https://user-images.githubusercontent.com/13084089/80928210-51cc1d00-8d9b-11ea-9e10-f1d2b4915ed7.png)
There's the entrypoint category on the left (doesn't have to be unique), class name in the middle and the interface it implements on the right.

Reply:
Ah, so that actually seems similar to how LiteLoader works. For my LiteLoader integration I just create an empty entry point, but that does seem like a nicer system. Seems like the "main class" utility could be pulled out into a separate function / class in a different way. I'll play with it a bit and see what feels good.

Reply:
I made some more changes which I think should support Fabric better. I removed `mainClass` as concept for all `ProjectConfig`s so you can do your own thing more easily with Fabric. The helper method in `BaseProjectCreator` can be used to create multiple classes, or not used at all if you like.